### ISSUE TITLE
Characterise the erasmusbip.org Ninja Tables endpoint

### ISSUE THREAD

**Issue by AntoniEntrepreneur** [AntoniEntrepreneur/BIPScraping#2]
## Question

Part of #1.

Does the Ninja Tables endpoint behind `list-of-bips` expose a **stable per-row identifier**, and how reliable is the endpoint as a long-lived data source?

The whole dedup design hangs on this. Right now the payload is pure field values with no id, and the obvious guess at a key (`Erasmus Code + title + start date`) collides 36 times across 636 rows.

Answer these:

1. **Row ids.** Can any parameter combination make `target_action=get-all-data` (or another `target_action`) return the underlying row id / post id? Ninja Tables 5.2.12, `provider: fluent-form`, `table_id=643`. Check the plugin source for the registered public actions and their accepted params rather than guessing at query strings.
2. **The `provider: fluent-form` angle.** The table is fed by Fluent Forms submissions. Does that imply an entry id reachable through some other public surface?
3. **Nonce lifetime.** `ninja_table_public_nonce` is scraped from the page HTML. Is it session-bound, IP-bound, or time-bound? How long does one stay valid? Does the request work at all without it? This decides whether every run must fetch the page first.
4. **Stability and etiquette.** Is the full-table response deterministic in row order? Any rate limiting, `robots.txt` restriction, or terms that bear on a daily automated fetch?
5. **Sort/filter params.** Does `default_sorting` (`old_first`) or any other param give a reliable "newest rows last" ordering we could lean on as a cheap recency signal?

Record findings as a Markdown file in the repo and link it here.

**Comment by AntoniEntrepreneur** [comment]
## No. There is no usable stable row id, and no parameter can produce one.

This is settled from plugin source, not from guessing at query strings. The row id exists server-side and is **deliberately stripped** before serialisation by a filter against admin-configured post meta that no request input touches. Stop looking for it.

Two things soften that, and one operational simplification falls out that is arguably the bigger win:

- **The 36 "collisions" on `Erasmus Code + title + start date` are not collisions.** They are the same real-world BIP re-submitted with corrections. The key is correctly *merging*, which is exactly the "changed item" case the map already wants.
- **Row order is deterministic and append-only** — `ORDER BY submissions.id ASC`. New rows always land last.
- **The nonce is never verified.** The page fetch can be deleted entirely.

Full write-up, with source citations and the live-test evidence: **[`docs/research/bip-endpoint.md`](https://github.com/AntoniEntrepreneur/BIPScraping/blob/research/bip-endpoint/docs/research/bip-endpoint.md)** (branch `research/bip-endpoint`).

### Sources

Ninja Tables **5.2.12 free** (exact version match to the live site) and Fluent Forms **6.2.13 free**, both read locally from the WordPress plugin repository. Pro adds admin features, not public routes. Plus ~15 live requests, spaced.

---

### 1. Row ids — **No**

`app/Hooks/actions.php:43` registers the handler under a doubled prefix (`wp_ajax_wp_ajax_ninja_tables_public_action`), which is why the bare action name 400s. `AjaxHandler::registerAjaxRoutes` then dispatches `target_action` against a whitelist of **exactly one entry**: `get-all-data`. There is no second route to find.

`FluentFormProvider::addEntryMeta` *does* build `id` (the `fluentform_submissions` PK) and `serial_number` — then throws them away:

```php
$defaultData = ['id' => $value->id, 'serial_number' => $value->serial_number, 'status' => $value->status];
return array_merge($value->user_inputs,
    array_intersect_key(array_merge($defaultData, ...), array_combine($columns, $columns)));
```

`$columns` is `get_post_meta($tableId, '_ninja_table_columns')` — the admin's column list. **No request parameter feeds it.** The only way to surface the id is for the site owner to add an "ID" column in wp-admin.

There is also an `___id___` row wrapper that would leak the id via a CSS class, but it is gated on `$dataProvider == 'default'` and this table is `fluent-form`, so it never runs.

(This also explains the earlier failures: `include_id` and `keys_only` aren't parameters at all.)

### 2. The fluent-form angle — **No**

Both namespaces are registered and discoverable, and both are permission-gated:

```
GET /wp-json/fluentform/v1/submissions?form_id=1  -> 401 rest_forbidden
GET /wp-json/ninjatables/v2/tables/643/item       -> 401 rest_forbidden
```

Ninja Tables reaches entries by short-circuiting `fluentform/verify_user_permission_fluentform_entries_viewer` for the duration of its own in-process query. Not reachable from outside.

### 3. Nonce — **Moot. It is never verified. Drop the page fetch.**

`getAllData` calls `ninjaTablesRequest()` (a plain `request->all()`) and validates nothing but that the post exists and is published. The only `wp_verify_nonce(..., 'ninja_table_public_nonce')` in the whole plugin is in an unrelated FluentCart handler.

| Request | Result |
| --- | --- |
| valid nonce | 200, 636 rows |
| **nonce omitted** | 200, 636 rows — byte-identical |
| **nonce = `deadbeef00`** | 200, 636 rows — byte-identical |

If it ever *did* start being enforced: it is time-bound only, **12–24h**, and neither session- nor IP-bound (the same value came back across different user agents with no cookies, because `wp_create_nonce` hashes uid `0` and an empty session token for anonymous visitors).

**Every run is one POST.** No HTML scraping, no regex against Elementor markup. Suggest still sending a placeholder nonce param as free insurance, so a future enforcement fails cleanly.

### 4. Stability and etiquette — **Deterministic; explicitly permitted**

Order chain is fully server-side: table post meta `old_first` -> `'ASC'` -> `EntryQuery::getResponses` -> `->orderBy('id', ...)`. A total order on an autoincrement PK, no tie-break ambiguity. Repeated full fetches minutes apart were byte-identical.

`robots.txt` **specifically carves the endpoint out** of the wp-admin disallow:

```
Disallow: /wp-admin/
Allow: /wp-admin/admin-ajax.php
```

No terms-of-service page; the privacy policy describes the table as a registry to be browsed and says nothing about automated access. No rate-limit headers seen, though Wordfence is installed — so excess volume would meet a WAF, not a 429. One POST/day at ~1.5 MB is a non-issue.

⚠️ **One genuine flag for the store-schema ticket:** rows carry personal data (~356 distinct coordinator emails, plus named individuals in free text). The map commits the seen-store **to the repo**. If the repo is or becomes public, that republishes a scraped, indexable copy under a GDPR controller, in a history that can't be redacted. Worth a decision: hash the emails, or keep the repo private.

### 5. Sort/filter — **`default_sorting` is a no-op, but yes, there's a recency signal**

`FluentFormProvider::data` accepts `$defaultSorting` and **never uses it**, calling `getOrderBy($tableId)` (post meta) instead. Confirmed: `old_first`, `new_first`, and omitting it all return byte-identical payloads.

The real param surface: `skip_rows`, `limit_rows` (cap 3000), `chunk_number`, `own_only`. Verified `limit_rows=5` == first 5 rows; `skip_rows=633&limit_rows=5` == last 3 rows.

**This explains `chunk_number=1` returning empty:** per-chunk is 3000, so that means `skip_rows=3000`, past the end of a 636-row table. And `per_page` was never a param — `limit_rows` is.

Because order is ascending by submission id, the table is **append-only: new rows always land last**, and `skip_rows=<count seen last run>` is a real incremental fetch. **Use it as a signal, not as the fetch strategy** — tail-only fetching can't see edits to existing rows, and one deleted submission shifts every offset, silently skipping a genuinely new row. Full fetch is free at this size.

---

### Identity: what to do instead

Inspecting all 36 collision groups: within a group, `description` differs in 22, deadline in 19, `description_4` in 16, status in 12 — while university differs in only 4. That's the signature of corrected re-submissions, not distinct BIPs:

- idx 145/146: `Università di Napoli Federico II` vs `University of Naples Federico II`
- idx 194/200: second submission adds the application-form link, fixes the deadline
- idx 120/134: status moved `Last Minute` -> `Full BIP - closed`

Two different BIPs sharing an Erasmus code *and* exact title *and* exact start date isn't plausible. And the 4 full-row duplicates sit at **211/212, 398/399, 497/498, 520/521** — all adjacent pairs, i.e. double-clicks on the form. Collapsing them is correct, not data loss.

**Recommended model** — split identity from content, which is what the missing id was obscuring:

- **Identity key:** normalised `Erasmus Code + title + start date` (NFKC + casefold + whitespace collapse) -> **595 entities over 636 rows**. Survives edits. Always computable: **zero** rows have any of the three fields empty.
- **Content hash:** SHA-256 over all 15 values, for change detection only. Never as identity — any edit mints a new hash, reporting one phantom disappearance plus one false "new".
- **Last row wins** within a group (ascending order ⇒ newest correction is live state).

⚠️ **Parsing warning:** the start-date field is not reliably a date. 633/636 are `DD-Mon-YY`; 3 contain spam junk (`NvUJstRdygaKBXJClx`, `uhFSPtjMKKwhbKcQnNmZ`, `oVlKYIgRXkNjQsEtOYXgEk`). The deadline field is **empty in 70 rows** and malformed in 3. Since the report filters on "deadline not passed", parse defensively and treat unparseable as unknown, not passed. The identity key must use the **raw** date string or those 3 rows lose their identity.

### What this changes in the design

1. **Delete the page-scrape step** from the BIP fetcher.
2. **Stop looking for an id** — ship the identity-key + content-hash model.
3. **The 36 collisions become the "changed item" path**, not a problem.
4. **Append-only row order answers the map's open "silent-failure detection" question** — a non-growing or shrinking row count is a ready-made liveness canary.
5. **Personal data in a committed store needs a decision** before the store schema lands.
